File: 2025_Oct/evelyn/week6/2025-11-24-binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans;
         if (!root) return ans;

        queue<pair<TreeNode*, int>> q;
        q.push({root, 0});

        while (!q.empty()) {
            auto [node, level] = q.front();
            q.pop();

            if (level==ans.size()) ans.push_back({});
            ans[level].push_back(node->val);

            if (node->left) q.push({node->left, level+1});
            if (node->right) q.push({node->right, level+1});

        }

        return ans;

    }
};

File: 2025_Oct/Soohyeun/2025-11-24-BinaryTreeLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        # bfs -> using queue (FIFO) -> put all children nodes in the queue in each levels
        # -> count number of current level nodes -> iterate over n times : explore current level nodes
        if not root:
            return []

        queue = deque([root])
        res = []

        while queue:
            number_nodes = len(queue)
            level_nodes = []

            for _ in range(number_nodes):
                curr_node = queue.popleft()
                level_nodes.append(curr_node.val)
                if curr_node.left:
                    queue.append(curr_node.left)
                if curr_node.right:
                    queue.append(curr_node.right)

            res.append(level_nodes)

        return res
